Base-Commit: 662ad1dc6900ffe1b1a2a6bc0fca01cba488d747
Base-Branch: 
diff --git a/download.go b/download.go
index 0777de8..4f31c30 100644
--- a/download.go
+++ b/download.go
@@ -409,3 +409,16 @@ func validateMedia(iv, file, macKey, mac []byte) error {
 	}
 	return nil
 }
+
+func (cli *Client) DownloadMediaWithUrl(ctx context.Context, url string, mediaKey []byte, appInfo MediaType, fileEncSha256 []byte, fileSha256 []byte) (data []byte, err error) {
+	return cli.downloadAndDecrypt(ctx, url, mediaKey, appInfo, fileEncSha256, fileSha256)
+}
+
+func GetMMSType(mediaType MediaType) string {
+	return mediaTypeToMMSType[mediaType]
+}
+
+type DownloadableMessageWithURL interface {
+	DownloadableMessage
+	GetUrl() string
+}
diff --git a/socket/framesocket.go b/socket/framesocket.go
index 37837c5..c046109 100644
--- a/socket/framesocket.go
+++ b/socket/framesocket.go
@@ -73,7 +73,7 @@ func (fs *FrameSocket) Close(code websocket.StatusCode) {
 	if code > 0 {
 		err := conn.Close(code, "")
 		if err != nil {
-			fs.log.Warnf("Error sending close to websocket: %v", err)
+			fs.log.Debugf("Error sending close to websocket: %v", err)
 		}
 	} else {
 		err := conn.CloseNow()
diff --git a/appstate/decode.go b/appstate/decode.go
--- a/appstate/decode.go
+++ b/appstate/decode.go
@@ -388,6 +388,15 @@ func (proc *Processor) DecodePatches(
 			if len(warn) > 0 {
 				proc.Log.Warnf("Warnings while updating hash for %s: %+v", list.Name, warn)
 			}
+			// nchat: log content of rejected patch for diagnostics
+			for i, mutation := range patch.GetMutations() {
+				indexMAC, _, index, syncAction, _, decErr := proc.decodeMutation(ctx, mutation, i, false)
+				if decErr != nil {
+					proc.Log.Warnf("Rejected %s patch v%d mutation #%d %s undecodable: %v", list.Name, patch.GetVersion().GetVersion(), i+1, mutation.GetOperation(), decErr)
+					continue
+				}
+				proc.Log.Warnf("Rejected %s patch v%d mutation #%d %s index %v mac %X ts %d value %v", list.Name, patch.GetVersion().GetVersion(), i+1, mutation.GetOperation(), index, indexMAC, syncAction.GetValue().GetTimestamp(), syncAction.GetValue())
+			}
 			return
 		}
 
